qmi/core/rpc.py

```diff
@@ -659,6 +676,26 @@ def make_rpc_forward_function(method_name: str):
         # Add non-blocking proxy.
         self.rpc_nonblocking = QMI_RpcNonBlockingProxy(context, descriptor)
 
+    def __enter__(self) -> "QMI_RpcProxy":
+        """The context manager definition is needed for the proxy as it will always be returned from QMI contexts,
```

Comment:
I think it's probably a good idea to put [this](https://docs.python.org/3/reference/datamodel.html#special-method-lookup) link in here for people who'd like to deep-dive.

Reply:
I still don't really get how that link relates with the Recursion error?

Reply:
It explains why we need `__enter__` and `__exit__` in the `QMI_RpcProxy` class. Normally, you expect that having it in `QMI_Instrument` would suffice, but because we return a `QMI_RpcProxy` we have to implement the context management protocol in `QMI_RpcProxy` (just as a stub) to satisfy the `with` keyword. The `with` keyword looks in the class to see if the context management protocol is implemented and not in the instance.

Reply:
with _actually executes_ the type(obj).__enter__ and type(obj).__exit__ methods.

Reply:
those class methods in turn call the instance methods, which are rpc methods